feat: add voice tracing and evaluation docs scaffold (#4368)
langchain-ai/docs: src/langsmith/trace-gemini-live.mdx

---
title: Trace Gemini Live applications
sidebarTitle: Gemini Live
description: Trace Gemini Live voice agents built with the Google Agent Development Kit (ADK) in LangSmith.
---

The [Gemini Live API](https://ai.google.dev/gemini-api/docs/live-api) enables low-latency, bidirectional voice interactions with Gemini models over a persistent WebSocket connection. This guide shows how to trace a Gemini Live voice agent built with the [Google Agent Development Kit (ADK)](https://google.github.io/adk-docs/streaming/) to LangSmith.

Gemini Live is a speech-to-speech model: it processes audio natively and exchanges a continuous stream of events with your application over a persistent WebSocket connection, rather than making discrete request/response calls. The sections below walk through those events and how to turn them into a LangSmith trace. For our high-level principles on getting the most out of your voice agent traces, see [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals).

## The ADK Live event model

> As the conversation runs, ADK streams a series of events to your application. Every event has the same shape, and most of its fields are optional — so you determine what an event represents from *which fields are populated*:

Author: ```suggestion
As the conversation runs, ADK streams a series of events to your application. These events rerpsent... Every event has the same shape, and most of its fields are optional — so you determine what an event represents from *which fields are populated*:
```

add a line about what the event is and represent and does

| Populated field | Meaning |
| --- | --- |
| `content.parts[*].inline_data` | A chunk of agent audio (PCM16 bytes). The agent's voice arrives as a flood of these. |
| `input_transcription` | A fragment of the *user's* speech transcript. A final event repeats the full utterance with `finished=True`. |
| `output_transcription` | A fragment of the *agent's* speech transcript. |
| `content.parts[*].function_call` | The model requested a tool (name and arguments). |
| `content.parts[*].function_response` | ADK executed the tool and is returning the result to the model. |
| `turn_complete` | The server finished its half of the exchange. |
| `interrupted` | The server detected user barge-in over the agent. Flush your speaker buffer. |

Two consequences shape how you trace these events:

> - **Fields can co-occur** on a single event (for example, audio plus a transcription fragment). Classify by checking fields in priority order — `interrupted` > `turn_complete` > transcriptions > function calls — rather than assuming one field per event.

Author: remove

> - **Transcription is opt-in.** You get no transcripts at all unless you enable them in the `RunConfig` (see [Step 1](#step-1-enable-transcription)). The `finished=True` transcription event carries the complete utterance, so there's no need to accumulate fragments client-side.

Author: add as a note later where relevant

> <Info>

Author: move to the next section as well

Audio rates differ by direction: ADK Live expects 16 kHz PCM16 input and produces 24 kHz output. If your microphone capture isn't 16 kHz, resample it on the send path.
</Info>

## How events map to LangSmith runs

To get the most out of your traces, capture each meaningful event and the data it contains in a single conversation trace, with one span per event:

```text
realtime_session                       ← root run (combined stereo conversation.wav; ls_modality="audio")
│   metadata: thread_id, model, event_count, duration_s
│
├─ input_transcription                 ← user speech fragment (payload → span inputs)
├─ output_transcription                ← agent speech fragment (payload → span outputs)
├─ function_call: get_weather          ← the model requested the tool
├─ function_response: get_weather      ← ADK ran the tool; result heading back
├─ turn_complete                       ← turn boundary
└─ interrupted                         ← barge-in
```

> A few conventions specific to ADK Live:

Author: remove this section

- **Skip the audio flood.** Events whose only payload is agent audio are played but never spanned — a few minutes of conversation produces thousands of them. Scrub everything you do record: replace audio `bytes` with a `"<N bytes>"` placeholder and truncate long strings.
- **Don't synthesize tool runs.** ADK executes the registered Python function *itself* between the `function_call` and `function_response` events — your client never runs the tool. Record those two events as the source of truth and synthesize nothing.
- **Root run and audio.** Stamp roll-up metadata (`event_count`, `duration_s`) and any stream error on the root run before finalizing it, and attach one combined recording — see [Record a single combined audio file](/langsmith/trace-voice-fundamentals#record-a-single-combined-audio-file). For the attachment API itself, see [Upload files with traces](/langsmith/upload-files-with-traces).

## Installation

```bash
pip install "google-adk>=2.0" google-genai "langsmith>=0.4"
```

Install `sounddevice` and `numpy` as well if you want to capture local audio and attach the conversation recording.

## Set up your environment

> This is the SDK tracing path — there is no OTLP or OpenTelemetry wiring.

Author: ```suggestion
The following steps demonstrate how to trace using our SDK. You can also trace using OpenTelemetry directly. See XYZ
```

export LANGSMITH_PROJECT=my-voice-app
export GOOGLE_API_KEY=...
```

## Quickstart

### Step 1: Enable transcription

Enable input and output transcription in the `RunConfig`, or there is nothing useful to trace:

```python
from google.adk.agents.run_config import RunConfig, StreamingMode
from google.genai import types as genai_types

run_config = RunConfig(
    response_modalities=["AUDIO"],
    streaming_mode=StreamingMode.BIDI,
    input_audio_transcription=genai_types.AudioTranscriptionConfig(),
    output_audio_transcription=genai_types.AudioTranscriptionConfig(),
)
```

### Step 2: Open the conversation root run

Open one run for the whole conversation and mark it as a voice trace with `ls_modality="audio"`, following the [single-trace convention](/langsmith/trace-voice-fundamentals#trace-each-conversation-as-a-single-trace). Keep this run open for the lifetime of the session and finalize it when the session ends.

```python
from langsmith import RunTree

session = RunTree(
    name="conversation",
    run_type="chain",
    extra={"metadata": {"thread_id": thread_id, "model": MODEL, "ls_modality": "audio"}},
)
session.post()
```

### Step 3: Trace each event

> For each event the model sends, skip the audio-only chunks (they're played to the speaker, not traced) and open a child span on the conversation root for everything else. Put the event's payload into the span — as an input for events from the user, or an output for events from the model:

Author: ```suggestion
For each event the model sends, open a child span on the conversation root to capture the event's payload. We recommend you put the even't payload into the psan as an input for events from the user, or output for events form the model. 

add note abotu skipping audio chunks bc its too noisy
```

---
title: Trace Gemini Live applications
sidebarTitle: Gemini Live
description: Trace Gemini Live voice agents built with the Google Agent Development Kit (ADK) in LangSmith.
---

The [Gemini Live API](https://ai.google.dev/gemini-api/docs/live-api) enables low-latency, bidirectional voice interactions with Gemini models over a persistent WebSocket connection. This guide shows how to trace a Gemini Live voice agent built with the [Google Agent Development Kit (ADK)](https://google.github.io/adk-docs/streaming/) to LangSmith.

> Gemini Live is a speech-to-speech model: it processes audio natively and exchanges a continuous stream of events with your application over a persistent WebSocket connection, rather than making discrete request/response calls. The sections below walk through those events and how to turn them into a LangSmith trace. For the conventions that apply to any voice agent, see [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals).

Author: For our high-level philosophy on getting the most out of your voice agent traces...

> Gemini Live is a speech-to-speech model: it processes audio natively and exchanges a continuous stream of events with your application over a persistent WebSocket connection, rather than making discrete request/response calls. The sections below walk through those events and how to turn them into a LangSmith trace. For our high-level philosophy on getting the most out of your voice agent traces, see [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals).

Author: ```suggestion
Gemini Live is a speech-to-speech model: it processes audio natively and exchanges a continuous stream of events with your application over a persistent WebSocket connection, rather than making discrete request/response calls. The sections below walk through those events and how to turn them into a LangSmith trace. For our high-level principles on getting the most out of your voice agent traces, see [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals).
```

## The ADK Live event model

> ADK Live has no event `type` tag. Every item yielded by `runner.run_live(...)` is the same class (`google.adk.events.Event`), and **every interesting field is optional**. The "kind" of an event is implicit in *which fields are populated*:

Author: woah. need an intro sentence on the fact that the model functions by sending events. then you can say the events have shape X. then you can say that most fields are optional and we can derive the type of event based on which fields are populated.

```
Every item yielded by `runner.run_live(...)` is the same class (`google.adk.events.Event`)
```

too nitty

| `content.parts[*].function_call` | The model requested a tool (name and arguments). |
| `content.parts[*].function_response` | ADK executed the tool and is returning the result to the model. |
| `turn_complete` | The server finished its half of the exchange. |
| `interrupted` | The server detected user barge-in over the agent. Flush your speaker buffer. |

Two consequences shape how you trace these events:

- **Fields can co-occur** on a single event (for example, audio plus a transcription fragment). Classify by checking fields in priority order — `interrupted` > `turn_complete` > transcriptions > function calls — rather than assuming one field per event.
- **Transcription is opt-in.** You get no transcripts at all unless you enable them in the `RunConfig` (see [Step 1](#step-1-enable-transcription)). The `finished=True` transcription event carries the complete utterance, so there's no need to accumulate fragments client-side.

<Info>
Audio rates differ by direction: ADK Live expects 16 kHz PCM16 input and produces 24 kHz output. If your microphone capture isn't 16 kHz, resample it on the send path.
</Info>

## Build the trace your way

> LangSmith doesn't ship a dedicated ADK SDK integration for the live path, so you build the trace yourself from the event stream. The upside is full control: you decide which events become spans, how they're named, and what lands in inputs versus outputs. This guide walks through an approach that works well, and you can adapt it to whatever shape you want.

Author: remove

## How events map to LangSmith runs

> One conversation is one trace, and each meaningful event becomes one span:

Author: To get the most out of your traces: you want to capture each meaningful event and the data that it contains in a single conversation trace.

## Installation

```bash
pip install "google-adk>=2.0" google-genai "langsmith>=0.4"
```

Install `sounddevice` and `numpy` as well if you want to capture local audio and attach the conversation recording.

## Set up your environment

This is the SDK tracing path — there is no OTLP or OpenTelemetry wiring.

```bash
export LANGSMITH_API_KEY=...
export LANGSMITH_TRACING=true
export LANGSMITH_PROJECT=my-voice-app
export GOOGLE_API_KEY=...
```

## Quickstart

### Step 1: Enable transcription

Enable input and output transcription in the `RunConfig`, or there is nothing useful to trace:

```python
from google.adk.agents.run_config import RunConfig, StreamingMode
from google.genai import types as genai_types

run_config = RunConfig(
    response_modalities=["AUDIO"],
    streaming_mode=StreamingMode.BIDI,
    input_audio_transcription=genai_types.AudioTranscriptionConfig(),
    output_audio_transcription=genai_types.AudioTranscriptionConfig(),
)
```

### Step 2: Open the conversation root run

> Start a root run for the whole conversation with `metadata={"ls_modality": "audio"}`, following the [single-trace convention](/langsmith/trace-voice-fundamentals#trace-each-conversation-as-a-single-trace). Keep this run open for the lifetime of the session and finalize it when the session ends.

Author: show code snippet for this please

### Step 3: Wrap each event in a span, skipping the audio flood

Factor the "should this event get a span?" decision into one place. Audio-only events are played but not traced; everything else opens a child span named after the event:

```python
from contextlib import nullcontext

def event_context(session, event):
    """Return a span for one event, or a no-op for the agent-audio flood."""
    if event.is_audio_only:
        return nullcontext()
    return session.event_span(

>         event.raw, session.now(), name=event.label, inbound=event.is_inbound

Author: only show the session code here. For this to make sense, you have to have created a session in one of the code samples above. Otherwise, people are like, "Well, what is session?" Also, don't worry too much about the specific implementation of ours. Other people are going to have their own implementations, and that's fine. What we need to show is what is the code for creating an event span and then what are we putting into that span. Okay.

    )
```

Wrap all event handling in that context. Play audio chunks as they arrive, flush the speaker on barge-in, and record the final (not fragment) transcripts:

```python
async for raw_event in runner.run_live(
    user_id=USER_ID,
    session_id=adk_session.id,
    live_request_queue=queue,
    run_config=run_config,
):
    event = LiveEvent(raw_event)

>     with event_context(session, event):

Author: Here, event context is a function that we've defined to make this easier to do. Is there a way that we can do that without defining our own function? Can we just inline directly, like with session.event.span, blah blah blah?

        if event.interrupted:
            speaker.clear()                      # barge-in: flush agent audio
        for chunk in event.audio_chunks:
            speaker.write(chunk)                 # played, not spanned
        if event.final_user_transcript:
            print("user :", event.final_user_transcript)
        if event.final_agent_transcript:
            print("agent:", event.final_agent_transcript)
```

> A small `LiveEvent` wrapper around the raw event keeps this loop declarative. Expose readable properties — `audio_chunks`, `final_user_transcript`, `final_agent_transcript`, `is_audio_only`, `label`, `is_inbound` — so the run loop isn't a nest of `if event.content and event.content.parts and ...` checks.

Author: This is great. Can we make it a note, and can we link directly to the file that contains the definition for live event?

### Step 4: Handle failure and finalize

> Wrap the consume loop so that a stream failure is stamped on the root run (`session.run.error = f"{type(exc).__name__}: {exc}"`) *and* triggers teardown — a dead `run_live` generator otherwise hangs the app silently. Finalize the session in a `finally` block so the root run always closes.

Author: remvoe

> ### Gotchas

Author: Troubleshooting?

Author: see how its written for livekit please

---
title: Trace Gemini Live applications
sidebarTitle: Gemini Live
description: Trace Gemini Live voice agents built with the Google Agent Development Kit (ADK) in LangSmith.
---

The [Gemini Live API](https://ai.google.dev/gemini-api/docs/live-api) enables low-latency, bidirectional voice interactions with Gemini models over a persistent WebSocket connection. This guide shows how to trace a Gemini Live voice agent built with the [Google Agent Development Kit (ADK)](https://google.github.io/adk-docs/streaming/) to LangSmith.

Before you start, read [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals). Gemini Live is a [speech-to-speech](/langsmith/trace-voice-fundamentals#speech-to-speech-s2s) architecture, so this guide applies the conventions for that shape: one conversation is one trace, you trace the **event payloads** that cross the wire (not turns), and you attach a single combined audio recording to the root run with `ls_modality="audio"`.

## Why you build the trace yourself

> ADK ships standard `gen_ai.*` OpenTelemetry instrumentation for its non-live execution paths (`Runner.run`, `BaseAgent.run_async`, `FunctionTool.run_async`, and so on). But `Runner.run_live` — the entire voice loop — is **not** instrumented. If you wire up OpenTelemetry and expect spans from the live path, you get a single empty root and nothing inside it.

Reviewer: ```suggestion
ADK ships standard `gen_ai.*` OpenTelemetry instrumentation for its non-live execution paths (`Runner.run`, `BaseAgent.run_async`, `FunctionTool.run_async`, and so on). But `Runner.run_live`, the entire voice loop, is **not** instrumented. If you wire up OpenTelemetry and expect spans from the live path, you get a single empty root and nothing inside it.
```

This is the same situation as the [OpenAI Realtime API](/langsmith/trace-openai-realtime): the framework hands you an event stream from a remote service, so you observe that stream and build the trace yourself with the LangSmith SDK. As a rule of thumb across voice frameworks:

> - **OpenTelemetry** — the framework runs the pipeline in-process and emits its own spans ([LiveKit](/langsmith/trace-with-livekit), [Pipecat](/langsmith/trace-with-pipecat)). You translate those spans.

Reviewer: ```suggestion
- **OpenTelemetry**: the framework runs the pipeline in-process and emits its own spans ([LiveKit](/langsmith/trace-with-livekit), [Pipecat](/langsmith/trace-with-pipecat)). You translate those spans.
```

> - **SDK** — the framework hands you an event stream from a remote service (OpenAI Realtime, Gemini Live via ADK). You build the trace.

Reviewer: ```suggestion
- **SDK**: the framework hands you an event stream from a remote service (OpenAI Realtime, Gemini Live via ADK). You build the trace.
```

## The ADK Live event model

> The biggest conceptual difference from other realtime APIs is that ADK Live has no event `type` tag. Every item yielded by `runner.run_live(...)` is the same class (`google.adk.events.Event`), and **every interesting field is optional**. The "kind" of an event is implicit in *which fields are populated*:

Reviewer: ```suggestion
The biggest conceptual difference from other realtime APIs is that ADK Live has no event `type` tag. Every item yielded by `runner.run_live(...)` is the same class (`google.adk.events.Event`), and **every interesting field is optional**. The "kind" of an event is implicit in **which fields are populated**:
```

| Populated field | Meaning |
| --- | --- |
| `content.parts[*].inline_data` | A chunk of agent audio (PCM16 bytes). The agent's voice arrives as a flood of these. |
| `input_transcription` | A fragment of the *user's* speech transcript. A final event repeats the full utterance with `finished=True`. |
| `output_transcription` | A fragment of the *agent's* speech transcript. |
| `content.parts[*].function_call` | The model requested a tool (name and arguments). |
| `content.parts[*].function_response` | ADK executed the tool and is returning the result to the model. |
| `turn_complete` | The server finished its half of the exchange. |
| `interrupted` | The server detected user barge-in over the agent. Flush your speaker buffer. |

Two consequences shape how you trace these events:

> - **Fields can co-occur** on a single event (for example, audio plus a transcription fragment). Classify by checking fields in priority order — `interrupted` > `turn_complete` > transcriptions > function calls — rather than assuming one field per event.

Reviewer: ```suggestion
```

> - **Transcription is opt-in.** You get no transcripts at all unless you enable them in the `RunConfig` (see [Step 1](#step-1-enable-transcription)). The `finished=True` transcription event carries the complete utterance, so there's no need to accumulate fragments client-side.

Reviewer: ```suggestion
```

<Info>
Audio rates differ by direction: ADK Live expects 16 kHz PCM16 input and produces 24 kHz output. If your microphone capture isn't 16 kHz, resample it on the send path.
</Info>

## How events map to LangSmith runs

> One conversation is one trace, and each meaningful event becomes one span:

Reviewer: ```suggestion
To get the most out of your traces: you want to capture each meaningful event and the data that it contains in a single conversation trace.
```

└─ interrupted                         ← barge-in
```

A few conventions specific to ADK Live:

- **Span names are derived, not given.** Because there's no `type` tag, derive the span name from the same field inspection: `interrupted`, `turn_complete`, `input_transcription`, `output_transcription`, `function_call: <name>`, `function_response: <name>`, with `event` as the fallback.
- **Direction.** Events carrying the user's transcription land in the span's `inputs`; everything the model or server sends back lands in `outputs`.
- **Skip the audio flood.** Events whose only payload is agent audio are played but never spanned — a few minutes of conversation produces thousands of them. Scrub everything you do record: replace audio `bytes` with a `"<N bytes>"` placeholder and truncate long strings.
- **Don't synthesize tool runs.** ADK executes the registered Python function *itself* between the `function_call` and `function_response` events — your client never runs the tool. Record those two events as the source of truth and synthesize nothing. (This contrasts with OpenAI Realtime, where the client runs the tool and gets a real nested tool run.)

> - **Root run and audio.** Stamp roll-up metadata (`event_count`, `duration_s`) and any stream error on the root run before finalizing it, and attach one combined recording — see [Record a single combined audio file](/langsmith/trace-voice-fundamentals#record-a-single-combined-audio-file).

Reviewer: ```suggestion
```

## Installation

```bash
pip install "google-adk>=2.0" google-genai "langsmith>=0.4"
```

Install `sounddevice` and `numpy` as well if you want to capture local audio and attach the conversation recording.

## Set up your environment

> This is the SDK tracing path — there is no OTLP or OpenTelemetry wiring.

Reviewer: ```suggestion
The following steps demonstrate how to trace using our SDK. You can also trace using OpenTelemetry directly.
```

```bash
export LANGSMITH_API_KEY=...
export LANGSMITH_TRACING=true
export LANGSMITH_PROJECT=my-voice-app
export GOOGLE_API_KEY=...
```

## Quickstart

### Step 1: Enable transcription

> Enable input and output transcription in the `RunConfig`, or there is nothing useful to trace:

Reviewer: ```suggestion
Enable input and output transcription in the `RunConfig`, or there is nothing useful to trace.

<Note>
**Transcription is opt-in.** You get no transcripts at all unless you enable them in the `RunConfig`. The `finished=True` transcription event carries the complete utterance, so there's no need to accumulate fragments client-side.
</Note>
```

```python
from contextlib import nullcontext

def event_context(session, event):
    """Return a span for one event, or a no-op for the agent-audio flood."""
    if event.is_audio_only:
        return nullcontext()
    return session.event_span(
        event.raw, session.now(), name=event.label, inbound=event.is_inbound
    )
```

Wrap all event handling in that context. Play audio chunks as they arrive, flush the speaker on barge-in, and record the final (not fragment) transcripts:

```python
async for raw_event in runner.run_live(
    user_id=USER_ID,
    session_id=adk_session.id,
    live_request_queue=queue,
    run_config=run_config,
):
    event = LiveEvent(raw_event)
    with event_context(session, event):
        if event.interrupted:
            speaker.clear()                      # barge-in: flush agent audio
        for chunk in event.audio_chunks:
            speaker.write(chunk)                 # played, not spanned
        if event.final_user_transcript:
            print("user :", event.final_user_transcript)
        if event.final_agent_transcript:
            print("agent:", event.final_agent_transcript)
```

> A small `LiveEvent` wrapper around the raw event keeps this loop declarative. Expose readable properties — `audio_chunks`, `final_user_transcript`, `final_agent_transcript`, `is_audio_only`, `label`, `is_inbound` — so the run loop isn't a nest of `if event.content and event.content.parts and ...` checks.

Reviewer: ```suggestion
<Note>
A small [`LiveEvent`]() wrapper around the raw event keeps this loop declarative. Expose readable properties (`audio_chunks`, `final_user_transcript`, `final_agent_transcript`, `is_audio_only`, `label`, `is_inbound`) so the run loop isn't a nest of `if event.content and event.content.parts and ...` checks.
</Note>
```

### Step 4: Handle failure and finalize

> Wrap the consume loop so that a stream failure is stamped on the root run (`session.run.error = f"{type(exc).__name__}: {exc}"`) *and* triggers teardown — a dead `run_live` generator otherwise hangs the app silently. Finalize the session in a `finally` block so the root run always closes.

Reviewer: ```suggestion
Wrap the consume loop so that a stream failure is stamped on the root run (`session.run.error = f"{type(exc).__name__}: {exc}"`) **and** triggers teardown. A dead `run_live` generator otherwise hangs the app silently. Finalize the session in a `finally` block so the root run always closes.
```

> ### Gotchas

Reviewer: Ah from your comment looks like some change wasn't carried over, but yes should rename

Reviewer: Troubleshooting, Caveats, Constraints

- **No transcription configs means empty-looking traces.** This is the most common failure mode. Both `input_audio_transcription` and `output_audio_transcription` must be set on the `RunConfig`.
- **Don't accumulate transcript fragments.** Use the `finished=True` event's full text; fragments are only for live UI display.
- **Don't span audio-only events.** A few minutes of conversation produces thousands of them.
- **Fields co-occur.** Classify by priority, not by assuming one field per event.

> - **Tools run inside ADK.** Don't synthesize your own tool runs — you'd double-count what `function_call` and `function_response` already record.

Reviewer: ```suggestion
- **Tools run inside ADK.** Don't synthesize your own tool runs. This would lead to double-count what `function_call` and `function_response` already record.
```

---
title: Trace Gemini Live applications
sidebarTitle: Gemini Live
description: Trace Gemini Live voice agents built with the Google Agent Development Kit (ADK) in LangSmith.
---

The [Gemini Live API](https://ai.google.dev/gemini-api/docs/live-api) enables low-latency, bidirectional voice interactions with Gemini models over a persistent WebSocket connection. This guide shows how to trace a Gemini Live voice agent built with the [Google Agent Development Kit (ADK)](https://google.github.io/adk-docs/streaming/) to LangSmith.

> Gemini Live is a speech-to-speech model: it processes audio natively and exchanges a continuous stream of events with your application over a persistent WebSocket connection, rather than making discrete request/response calls. The sections below walk through those events and how to turn them into a LangSmith trace. For our high-level principles on getting the most out of your voice agent traces, see [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals).

Reviewer: ```suggestion
Gemini Live is a speech-to-speech model: it processes audio natively and exchanges a continuous stream of events with your application over a persistent WebSocket connection, rather than making discrete request/response calls. The following sections show those events and how to turn them into a LangSmith trace. For our high-level principles on getting the most out of your voice agent traces, see [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals).
```

Removing directional language for accessibility/screenreaders

## The ADK Live event model

> As the conversation runs, ADK streams a series of events to your application. Each event reports something that happened in the conversation: a chunk of audio, a transcript fragment, a tool call, a turn boundary, or an interruption. Every event has the same shape, and most of its fields are optional, so you determine what an event represents from *which fields are populated*:

Reviewer: ```suggestion
As the conversation runs, ADK streams a series of events to your application. Each event reports something that happened in the conversation: a chunk of audio, a transcript fragment, a tool call, a turn boundary, or an interruption. Every event has the same shape, and most of its fields are optional, so you determine what an event represents from **which fields are populated**:
```

---
title: Trace Gemini Live applications
sidebarTitle: Gemini Live
description: Trace voice agents built on the Gemini Live API in LangSmith.
---

<Info>
**Draft structure for review.** Code samples, the final message mapping, and screenshots are placeholders to be filled in.

> </Info>

Author: remove

Before you start, read [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals) to understand the three core conventions this guide applies:

- Attach conversation audio to the root run.
- Trace each conversation as a single trace, not one trace per turn.
- Mark the root run with `ls_modality="audio"`.

## How it works

Like other realtime APIs, Gemini Live streams messages over a session connection rather than discrete request/response calls. To trace a conversation, open a single trace when the session starts, then nest each turn (and its tool calls) underneath as the session streams messages.

| Gemini Live message | Trace mapping |
| --- | --- |
| Session setup | Start the root run; record session config. |
| User audio / input transcription | Capture a user turn; attach input audio and transcript. |
| Model turn / output transcription | Start an assistant turn; attach output audio and transcript. |
| Tool / function calls | Nest a tool run within the turn. |

> | Usage metadata | Record token counts and latency on the turn. |

Author: We don't want to capture things as turns, but rather capture the payloads of the events that are sent from/to the model server. Can you look at the gemini live docs to get a list of what those events are with a quick summary (+ link to the docs).
